specifications: finalization of package repositories spec (#3333)
snapcore/snapcraft: specifications/package-repositories.org

- *Description:* List of deb types to enable.

- *Default:* If unspecified, format is assumed to be =deb=, i.e. =[deb]=.

- *Examples:*

  - =formats: [deb]=

  - =formats: [deb, deb-src]=

** Property: components (required if specifying suites)

- *Type:* list[string]

- *Description:* Apt repository components to enable: e.g. =main=, =multiverse=, =unstable=.

- *Examples:*

  - =components: [main]=

  - =components: [main, multiverse, universe, restricted]=

> ** Property: key-id (required if not using ppa)

Author: ```suggestion
** Property: key-id (required if not using the ppa type)
```

  - *Type:* string

  - *Description:* GPG key identifier.

>     - Sometimes referred to as "thumbprint" or "fingerprint".

Author: ```suggestion
    - Sometimes referred to as a "thumbprint" or "fingerprint".
```

>     - Must be 40 characters (full form).

Author: ```suggestion
    - Must be 40 characters (the long-form thumbprint).
```

  - *Notes:*

>     - If not using key-server, Snapcraft will look for corresponding key at:

Author: ```suggestion
    - If not using the key-server, Snapcraft will look for corresponding key at:
```

      =<project>/snap/keys/<key-id[0:8]>.asc=.

    - To determine a key-id from a given key file using =gpg=:
      =gpg --import-options show-only --import <file>=

  - *Format:* alphanumeric, dash =-=, and underscores =_= permitted.

  - *Examples:*

    - =key-id: 590CA3D8E4826565BE3200526A634116E00F4C82=

>       Snapcraft will install corresponding key at

Author: ```suggestion
      Snapcraft will install a corresponding key at
```

      =<project>/snap/keys/590CA3D8.key=.

** Property: key-server (optional)

  - *Type:* string

  - *Description:* Key server to fetch key =<key-id>= from.

  - *Default:* If unspecified, Snapcraft will attempt to fetch a specified key
    from keyserver.ubuntu.com.

  - *Format:* Key server URL supported by =gpg --keyserver=.

  - *Examples:*

    - =key-server: keyserver.ubuntu.com=

    - =key-server: hkp://keyserver.ubuntu.com:80=

** Property: path (required if not using suites & components)

  - *Type:* string

>   - *Description:* Absolute path to repository from URL.

Author: ```suggestion
  - *Description:* Absolute path to repository (from =url=).
```

    Apt sources using absolute paths are incompatible with suites and components.

  - *Format:* Path starting with =/=.

  - *Examples:*

    - =path: /=

    - =path: /my-repo=

** Property: suites (required if not using path)

  - *Type:* string

  - *Description:* Repository suites to enable.

  - *Notes:*

>     - If the deb URL does not look like it has a suite defined, it is likely

Author: ```suggestion
    - If your deb URL does not look like it has a suite defined, it is likely
```

  - type: apt
    components: [main]
    suites: [xenial]
    key-id: 78E1918602959B9C59103100F1831DDAFC42E99D
    url: http://ppa.launchpad.net/snappy-dev/snapcraft-daily/ubuntu

  # Apt repository enabling deb sources.
  - type: apt
    formats: [deb, deb-src]
    components: [main]
    suites: [xenial]
    key-id: 78E1918602959B9C59103100F1831DDAFC42E99D
    url: http://ppa.launchpad.net/snappy-dev/snapcraft-daily/ubuntu

  # Absolute path repository with implied "/".
  - type: apt
    key-id: AE09FE4BBD223A84B2CCFCE3F60F4B3D7FA2AF80
    url: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64`

  # Absolute path repository with explicit path and formats.
  - type: apt
    formats: [deb]
    path: /
    key-id: AE09FE4BBD223A84B2CCFCE3F60F4B3D7FA2AF80
    url: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64`
#+END_SRC

* Key search methodology

1. =<key-id>= will be checked at =<project>/snap/keys/<key-id[0:8]>.asc=.

   - If =<key-id[0:8]>.asc= file exists and =key-id= matches, the key file will
     be imported.

   - If =<key-id[0:8]>.asc= file exists and =key-id= does not match, Snapcraft

>      will error.

Reviewer: ```suggestion
     must error.
```

   - If =<key-id[0:8]>.asc= file does not exist, continue to step 2.

> 2. =<key-id>= will be queried from =<key-server>=, defaulting to =keyserver.ubuntu.com=.

Author: ```suggestion
2. =<key-id>= will be queried from the =<key-server>=, defaulting to =keyserver.ubuntu.com=.
```

   - If key is found, the key will be imported.

   - If key not found, an error will be presented to the user:

#+BEGIN_SRC python
Failed to install GPG key: GPG key {key_id} not found on key server {key_server!r}.

Recommended resolution:
Verify GPG key ID and key server are correct, or install key to
<project>/snap/keys/{key_id[0:8]}.asc.

Detailed information:
GPG key ID: <key-id>
GPG key server: <key-server>
#+END_SRC

> Once all repositories have been processed, if there are any unused keys present

Author: ```suggestion
If there are any unused keys present in =<project>/snap/keys=, Snapcraft must error.
```

> in =<project>/snap/keys=, Snapcraft will error.

Reviewer: ```suggestion
in =<project>/snap/keys=, Snapcraft must error.
```

> in =<project>/snap/keys=, Snapcraft must error.

Author: ```suggestion
```
